subsys/lorawan/lorawan.c

```diff
@@ -0,0 +1,540 @@
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_NWK_S_ENC_KEY;
+	mib_req.Param.NwkSEncKey = join_cfg->abp.nwk_skey;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_APP_S_KEY;
+	mib_req.Param.AppSKey = join_cfg->abp.app_skey;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	mib_req.Type = MIB_NETWORK_ACTIVATION;
+	mib_req.Param.NetworkActivation = ACTIVATION_TYPE_ABP;
+	LoRaMacMibSetRequestConfirm(&mib_req);
+
+	return LORAMAC_STATUS_OK;
+}
+
+int lorawan_join(const struct lorawan_join_config *join_cfg)
+{
+	LoRaMacStatus_t status;
+	int ret;
+
+	k_mutex_lock(&lorawan_join_mutex, K_FOREVER);
+
+	if (join_cfg->mode == LORAWAN_ACT_OTAA) {
+		join_status = false;
+		status = lorawan_join_otaa(join_cfg);
+		if (status != LORAMAC_STATUS_OK) {
+			LOG_ERR("OTAA join failed: %s",
+				log_strdup(status2str(status)));
+			ret = -EINVAL;
+			goto out;
+		}
+
+		LOG_DBG("Network join request sent!");
+
+		/*
+		 * We can be sure that the semaphore will be released for
+		 * both success and failure cases after a specific time period.
+		 * So we can use K_FOREVER and no need to check the return val.
+		 */
+		k_sem_take(&lorawan_config_sem, K_FOREVER);
+	} else if (join_cfg->mode == LORAWAN_ACT_ABP) {
+		status = lorawan_join_abp(join_cfg);
+		if (status != LORAMAC_STATUS_OK) {
+			LOG_ERR("ABP join failed: %s",
+				log_strdup(status2str(status)));
+			ret = -EINVAL;
+			goto out;
+		}
+	} else {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	if (join_status) {
+		ret = 0;
+	} else {
+		/* TODO: Return the exact error code */
+		ret = -EINVAL;
+	}
+
+out:
+	k_mutex_unlock(&lorawan_join_mutex);
+	return ret;
+}
+
+int lorawan_set_class(enum lorawan_class dev_class)
+{
+	LoRaMacStatus_t status;
+	MibRequestConfirm_t mib_req;
+
+	mib_req.Type = MIB_DEVICE_CLASS;
+
+	switch (dev_class) {
+	case LORAWAN_CLASS_A:
+		mib_req.Param.Class = CLASS_A;
+		break;
+	case LORAWAN_CLASS_C:
+		mib_req.Param.Class = CLASS_C;
+		break;
+	default:
+		return -EINVAL;
+	};
+
+	status = LoRaMacMibSetRequestConfirm(&mib_req);
+	if (status != LORAMAC_STATUS_OK) {
+		LOG_ERR("Failed to set device class: %s",
+			status2str(status));
+		return mac_status_to_errno[status];
+	}
+
+	return 0;
+}
+
+int lorawan_set_datarate(enum lorawan_datarate dr)
+{
+	/* Bail out if using ADR */
+	if (lorawan_adr_enable) {
+		return -EINVAL;
+	}
+
+	lorawan_datarate = dr;
+
+	return 0;
+}
+
+void lorawan_enable_adr(bool enable)
+{
+	MibRequestConfirm_t mib_req;
+
+	if (enable != lorawan_adr_enable) {
+		lorawan_adr_enable = enable;
+
+		mib_req.Type = MIB_ADR;
+		mib_req.Param.AdrEnable = lorawan_adr_enable;
+		LoRaMacMibSetRequestConfirm(&mib_req);
+	}
+}
+
+int lorawan_set_conf_msg_tries(uint8_t tries)
+{
+	lorawan_conf_msg_tries = tries;
+
+	return 0;
+}
+
+int lorawan_send(uint8_t port, uint8_t *data, uint8_t len, uint8_t flags)
+{
+	LoRaMacStatus_t status;
+	McpsReq_t mcpsReq;
+	LoRaMacTxInfo_t txInfo;
+	int ret = 0;
+	bool empty_frame = false;
+
```

Comment:
If we are going to have both a confirmed and an unconfirmed flag, we need to check whether the flag combination is valid here. For example, what happens if we both confirmed and unconfirmed are set? What if none is set?

I would argue that given the properties of the protocol, unconfirmed transmissions should probably be the default mode and anything else needs to be specified by a flag.